### PR TITLE
add kernel rom procedure table

### PR DESCRIPTION
**Pull request by grjte**
This PR completes the building of the auxiliary trace columns for the kernel ROM by building the `p1` column as described [here](https://github.com/0xPolygonMiden/miden-vm/issues/231). Since this reuses the column used for the hasher's sibling table, this includes a refactor of the chiplets auxiliary trace handling.

- [x] refactor chiplets to have a single auxiliary trace builder for both the shared chiplets bus and the virtual table, which is now shared (formerly the hasher sibling table)
- [x] add kernel rom procedures to the chiplets virtual table
- [x] update existing kernel rom tests to check the virtual table

# Notes
- doc updates will be handled in a future PR
- we no longer have a boundary constraint enforcing the correctness of the hasher sibling table, so we should decide how to replace this. We can enforce that the auxiliary column equals ONE for all cycles between the end of the hasher trace and the beginning of the kernel ROM trace, but I'm not sure if you had something else in mind
- I'm not fully convinced by this design for building the auxiliary trace, but we've discussed revisiting the entire pattern in the future, so I tried to stick fairly closely to the previous design and leave a rethink of this for another time.

### CODE REVIEW

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left a few small comments inline. These are mostly related to using slightly less generic names for what used to be `SiblingTable`. The best alternative I could come up with was "ChipletsVirtualTable" (or `ChipletsVTable`) but maybe there is something better.

Also, I think one thing missing from this PR (or it could be done in a follow-up PR as well) is updates to boundary constraints. Specifically, we need to enforce that chiplets vtable last value is not `ONE` but is equivalent to the kernel ROM table.

I think there are two parts to this:
1. In the `air` crate, we should add a boundary constraint for chiplet auxiliary columns (I don't think we do this right now). For the bus column, first/last value should be `ONE`, and for vtable column, first value should be ONE and last value should depend on the Kernel.
2. We might want to add a check in the prover in the `get_pub_inputs()` method. Though, I'm not sure we can enforce such checks against auxiliary columns yet as `get_pub_inputs()` does not receive randomness required to reduce public input tables to single values. This might be something that needs to be updated in Winterfell first.

**Comment by grjte** [comment]
> Also, I think one thing missing from this PR (or it could be done in a follow-up PR as well) is updates to boundary constraints. Specifically, we need to enforce that chiplets vtable last value is not `ONE` but is equivalent to the kernel ROM table.
> 
> I think there are two parts to this:
> 
>     1. In the `air` crate, we should add a boundary constraint for chiplet auxiliary columns (I don't think we do this right now). For the bus column, first/last value should be `ONE`, and for vtable column, first value should be ONE and last value should depend on the Kernel.
> 
>     2. We might want to add a check in the prover in the `get_pub_inputs()` method. Though, I'm not sure we can enforce such checks against auxiliary columns yet as `get_pub_inputs()` does not receive randomness required to reduce public input tables to single values. This might be something that needs to be updated in Winterfell first.

I tried to address the fact that we require changes to the boundary constraints in my notes in the PR description, although I didn't specifically mention the new value at the last row, instead focusing on the fact that we'll need to handle the constraint for the end of the sibling table, which can no longer be enforced by a boundary constraint at the last row. We didn't previously have these boundary constraints in the AIR, which is why I didn't put them there. In my mind, these were going to be updated in the documentation (in the next PR) and then subsequently updated in the AirScript files.

The new boundary constraint at the last boundary is easy to update & describe, but it would be good to get confirmation on how you'd like to handle the sibling table constraint that I asked about in the description.

Regarding (1) specifically - do you want this added to the `air` crate instead of in the docs + AirScript as I had assumed? Adding only the new boundary constraint without taking care of the enforcing the sibling table seems like a job half-done, which is why I hadn't done this yet.
Regarding (2) - yes, I'm not sure how we could do this now

File: processor/src/chiplets/aux_trace/mod.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+/// Contains all relevant information and describes how to construct execution trace of hasher-
+/// related auxiliary trace columns (used in multiset checks).
+///
+/// Currently, this manages construction of a single column representing the state of the sibling
+/// table (used in Merkle root update computation).
+#[derive(Debug, Clone, Default)]
+pub struct TableTraceBuilder {
+    pub(super) hints: Vec<(u32, ChipletsTableUpdate)>,
+    pub(super) rows: Vec<ChipletsTableRow>,
+}
```
[the remark this answers is not in the document]
Comment (grjte) [reply processor/src/chiplets/aux_trace/mod.rs:141]: Yeah, the only reason I didn't do something like this in the first place is that it's really long. I've changed it though

File: processor/src/chiplets/aux_trace/mod.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+impl AuxColumnBuilder<ChipletsTableUpdate, ChipletsTableRow, u32> for TableTraceBuilder {
+    /// Returns a list of rows which were added to and then removed from the sibling table.
+    ///
+    /// The order of the rows in the list is the same as the order in which the rows were added to
+    /// the table.
+    fn get_table_rows(&self) -> &[ChipletsTableRow] {
+        &self.rows
+    }
```
[the remark this answers is not in the document]
Comment (grjte) [reply processor/src/chiplets/aux_trace/mod.rs]: Thanks for catching the comments - I believe I've updated them all now.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

> In my mind, these were going to be updated in the documentation (in the next PR) and then subsequently updated in the AirScript files.

Yep - makes sense. Let's leave this until AirScript output is incorporated into the prover/verifier (so, no need to update the `air` crate now). I was just thinking if we can make the verifier fail somehow if the wrong Kernel is provided - but I guess without actually enforcing these constraints, we can't really do much.

> The new boundary constraint at the last boundary is easy to update & describe, but it would be good to get confirmation on how you'd like to handle the sibling table constraint that I asked about in the description.

We could do this by predicating the constraint on the value of of the first selector column of the chiplets table. Specifically, this constraint would be enforced only when values in this column switch from $0$ to $1$. Basically, we'd use the expression $s_0' - s_0$ as a selector.

It is also possible that we don't need to enforce this at all - but this requires more thinking, and for now, I think we can stay on the safe side for now.

> I'm not fully convinced by this design for building the auxiliary trace, but we've discussed revisiting the entire pattern in the future, so I tried to stick fairly closely to the previous design and leave a rethink of this for another time.

Agreed! One way I was thinking of is to have a simplified version of `AuxColumnBuilder` trait. This trait would have a `build_row_values()` function with basically the same signature as we have now. The main difference would be that instead of relying on a complex hint etc. structure, we'd get all the data needed for computing row values (and inverse row values) from the `main_trace`.

The formulas for some columns may get rather complicated - but the whole logic would be self-contained in one place, and if we use constants for column offsets, it would probably be rather maintainable.

We may lose out on performance a bit - but my guess is not enough to matter (though, would be good to confirm).

Also, this is just an idea, and there could be better approaches too.

**Pull request merged**